FILE_NAME  marked as obsolete because it does not support Unicode

jvelilla opened the pull request:
and  PATH should be used instead.
Added new creations procedures make_temporary_open and
make_temporary_open_with_prefixshould be added to the class  FILE  to create and open a temporary file

jvelilla committed:
FILE_NAME  marked as obsolete because it does not support Unicode
and  PATH should be is used instead.
Added new creation procedures make_temporary_open and
make_temporary_open_with_prefixshould be added to the class  FILE  to create and open a temporary file

jocelyn commented on Src/library/base/elks/kernel/file.e:
```diff
@@ -171,6 +171,39 @@ feature -- Initialization
+			l_file_info: FILE_INFO
+			s: STRING
+		do
+				-- UX platforms
```
Is this comment expected here?

jvelilla replied:
No, removed.

jocelyn commented on Src/library/base/elks/kernel/file.e:
```diff
@@ -171,6 +171,39 @@ feature -- Initialization
+				-- UX platforms
+				-- http://man7.org/linux/man-pages/man3/mkstemp.3.html
+			create l_temp.make (a_prefix + "XXXXXX")
+			l_res := eif_temp_file (l_temp.item)
```
I would rename `l_res`  as `fd`  for file descriptor

jvelilla replied:
updated as `l_fd`

jocelyn replied:
I was suggesting `fd` as it is already used in similar code, and as it is short  (2 letters), it does not need the `l_` prefix.
Far from being critical.

jocelyn commented on Src/library/base/elks/kernel/file.e:
```diff
@@ -2093,6 +2126,120 @@ feature {NONE} -- Implementation
+				    return -1;
+				  }
+
+				    random_time_bits = (((unsigned long long)ftNow.dwHighDateTime << 32)
```
there are a few wrong indentations.
Please reindent correctly, and there are a few empty lines that could be removed ... to have smaller body.

jvelilla replied:
updated.

jocelyn commented on Src/library/base/elks/kernel/file_name.e:
```diff
@@ -9,6 +9,8 @@ note
 class
 	FILE_NAME
 
+obsolete "This class is obsolete, use PATH instead"
```
It is missing a date ... I guess the compiler reported this missing date as warning.

jvelilla replied:
fixed

jvelilla replied:
added date

jocelyn commented on Src/library/base/elks/kernel/file_name.e:
```diff
@@ -24,6 +26,11 @@ feature {NONE} -- Initialization
 	make_temporary_name
 			-- Create a temporary filename.
+		note
+				EIS:"name={FILE}.make_temporary_open", "src=eiffel:?class=FILE&feature=make_temporary_open","protocol=uri"
```
bad indentations

jvelilla replied:
fixed.

jocelyn commented on Src/library/base/elks/kernel/file_name.e:
```diff
@@ -24,6 +26,11 @@ feature {NONE} -- Initialization
+				EIS:"name={FILE}.make_temporary_open", "src=eiffel:?class=FILE&feature=make_temporary_open","protocol=uri"
+				EIS:"name={FILE}.make_temporary_open_with_prefix", "src=eiffel:?class=FILE&feature=make_temporary_open_with_prefix","protocol=uri"
+		obsolete
+				"Use `{FILE}.make_temporary_open or {FILE}.make_temporary_open_with_prefix' instead. [2019-05-31]"
```
you should use the new convention with two back-ticks `foo` instead of back-tick and quote `foo'

jocelyn replied:
Is there a way to implement FILE_NAME.make_temporary_name without tempnam ?
If yes, even if this is not 100% safe, it would be good to have it.

Otherwise we'll still have the tempnam C warning for a few releases.

jvelilla replied:
Need to be checked, but that's the idea to make it obsolete.

kwaxer commented on Src/library/base/elks/kernel/file.e:
```diff
@@ -171,6 +171,39 @@ feature -- Initialization
 			open_append: is_open_append
 		end
 
+	make_temporary_open
```
I'd better use `make_open_temporary` to use the same convention as `make_open_read`.

jvelilla replied:
updated.

kwaxer commented on Src/library/base/elks/kernel/file.e:
```diff
@@ -171,6 +171,39 @@ feature -- Initialization
+		ensure
+			exists: exists
+			open_read: is_open_read
+			open_append: is_open_write
```
The tag name does not match the permission.

jvelilla replied:
updated.

kwaxer commented on Src/library/base/elks/kernel/file.e:
```diff
@@ -171,6 +171,39 @@ feature -- Initialization
+			open_append: is_open_write
+		end
+
+	make_temporary_open_with_prefix (a_prefix: READABLE_STRING_GENERAL)
```
Same here: `make_open_temporary_with_prefix`.

jvelilla replied:
updated.

kwaxer commented on Src/library/base/elks/kernel/file.e:
```diff
@@ -171,6 +171,39 @@ feature -- Initialization
+		ensure
+			exists: exists
+			open_read: is_open_read
+			open_append: is_open_write
```
See above.

jvelilla replied:
updated.

kwaxer commented on Src/library/base/elks/kernel/file.e:
```diff
@@ -2093,6 +2126,120 @@ feature {NONE} -- Implementation
 
+	eif_temp_file (a_name_template: POINTER): INTEGER
+			-- Create a unique temporary file name and return a file descriptor.
+			--| UX platforms: The `mkstemp()' function generates a unique temporary filename from
```
UX is used for "User Experience", not for "Unix".

kwaxer replied:
I'd move this comment inside the inline C code.

jvelilla replied:
Oh yes got confused with *nix, ok.

jvelilla committed:
Fixed comments and bad identations.

kwaxer commented on Src/library/base/elks/kernel/file.e:
```diff
@@ -2093,6 +2126,120 @@ feature {NONE} -- Implementation
+		alias
+			"[
+			#ifdef EIF_WINDOWS
+				#include <fcntl.h>
```
I's not OK to put include inside C code. It should be in "external" part.

kwaxer replied:
According to https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/open-wopen?view=vs-2017 the required header is `<io.h>`.

jvelilla replied:
Can we add more than one header in the "external" if needed?, It was part of the original code so I keep it.

jvelilla replied:
I guess why the original code includes the header <fcntl.h> it defines 
O_RDWR  _O_RDWR
 O_CREAT  _O_CREAT
O_EXCL     _O_EXCEL
O_BINARY _O_BINARY
So  maybe it's to increase portability

kwaxer commented on Src/library/base/elks/kernel/file.e:
```diff
@@ -2093,6 +2126,120 @@ feature {NONE} -- Implementation
+			#ifdef EIF_WINDOWS
+				#include <fcntl.h>
+				#ifdef _MSC_VER
+				#define open _open
```
I believe these defines are not needed, and `_open` and `_close` can be used directly.

kwaxer commented on Src/library/base/elks/kernel/file.e:
```diff
@@ -2093,6 +2126,120 @@ feature {NONE} -- Implementation
+				#ifdef _MSC_VER
+				#define open _open
+				#define close _close
+				#define EINVAL 22 /* Invalid argument  */
```
These constants are already defined. Have a look at https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/open-wopen?view=vs-2017

jvelilla replied:
Yes, I just use the [original code](https://github.com/Kazade/Tasks/blob/master/u1dbrepo/src/mkstemp_compat.c),  maybe it's there for compatibility reasons, but the code is not commented. I will check.

jvelilla replied:
Updated

kwaxer commented on Src/library/base/elks/kernel/file.e:
```diff
@@ -2093,6 +2126,120 @@ feature {NONE} -- Implementation
+				  return -1;
+				#endif // _MSC_VER
+			#else
+				return mkstemp ($a_name_template);
```
What about FILE.name? What does it return?

jvelilla replied:
Not sure to understand, in that case, we are calling the Unix mkstemp method.

jocelyn replied:
I think Alexander is asking what will be the result of `FILE.name` ? Does it return the expected temporary name?

jvelilla replied:
Yes, it returns the expected temporary name.

kwaxer commented on Src/library/base/elks/kernel/file.e:
```diff
@@ -2093,6 +2126,120 @@ feature {NONE} -- Implementation
+				      v /= 62;
+				      XXXXXX[5] = letters[v % 62];
+
+				      fd = open (tmpl, O_RDWR | O_CREAT | O_EXCL | O_BINARY, _S_IREAD | _S_IWRITE);
```
Should the file name be recorded somewhere, so that FILE.name returns meaningful reasult?

jvelilla replied:
Not sure to understand, the function returns a file descriptor and a name using the template pattern if everything is ok or an error.

kwaxer commented on Src/library/base/ise/kernel/dotnet/file.e:
```diff
@@ -165,6 +165,22 @@ feature -- Initialization
+			-- Create a file object with a unique temporary file name,
+			-- with read/write mode.
+		local
+			l_temp: STRING
```
Shouldn't it be `STRING_32`?

jvelilla replied:
Yes.

kwaxer commented on Src/library/base/ise/kernel/dotnet/file.e:
```diff
@@ -165,6 +165,22 @@ feature -- Initialization
 			open_append: is_open_append
 		end
 
+	make_temporary_open
```
What about the version `..._with_prefix`?

jvelilla replied:
Added

jvelilla committed:
Reafactored creation procedures name to follow existing convetions
Updated postconditiong tags to match the expected mode.
Removed unneded constants in the C external code,
removed the include form the body to the external header.

For .Net updated names to follow existing convetions and added a new
creation procedure to create temporary file with prefix.

jvelilla commented:
Closed, committed on rev=102700
